### PR TITLE
Ruby Enumerable Solo Project

### PR DESCRIPTION
**Pull request by camfilho**
## Hello!

This is the Ruby's solo project that can be found [here](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#project-2-enumerable-methods) <br>
***
The main file is called "**my_enum.rb**" and the tests are located in spec/enumerable_spec.rb.

Any question, you can find me on Slack "Carlos Augusto de Medeiros Filho"

### CODE REVIEW

File: my_enum.rb [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+      my_each do |i|
+        return false unless yield(i)
+      end
+    else
+      my_each do |i|
+        return false unless i
+      end
+    end
+    true
+  end
```
Comment (maelfosso) [inline my_enum.rb]:
```ruby
irb(main):008:0> array
=> [6, 5, 0, 7, 8, 7, 8, 6, 4, 0, 5, 1, 7, 6, 4, 4, 7, 2, 1, 0, 2, 4, 8, 8, 0, 7, 3, 8, 2, 2, 2, 1, 2, 8, 4, 0, 1, 7, 5, 3, 2, 2, 7, 3, 0, 6, 5, 7, 3, 5, 5, 6, 3, 8, 1, 0, 2, 4, 2, 1, 7, 5, 5, 2, 4, 4, 2, 3, 5, 1, 6, 5, 6, 2, 1, 8, 5, 7, 0, 4, 6, 3, 0, 5, 6, 8, 7, 7, 5, 0, 2, 0, 5, 0, 6, 5, 0, 6, 8, 4]
irb(main):009:0> array.all?(Integer)
=> true
irb(main):010:0> array.my_all?(Integer)
Traceback (most recent call last):
        5: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        4: from (irb):10
        3: from my_enum.rb:64:in `my_all?'
        2: from my_enum.rb:20:in `my_each'
        1: from my_enum.rb:65:in `block in my_all?'
NoMethodError (undefined method `match?' for 6:Integer)
irb(main):011:0> 
```

Comment (maelfosso) [reply my_enum.rb]:
```ruby
irb(main):013:0> array
=> [6, 5, 0, 7, 8, 7, 8, 6, 4, 0, 5, 1, 7, 6, 4, 4, 7, 2, 1, 0, 2, 4, 8, 8, 0, 7, 3, 8, 2, 2, 2, 1, 2, 8, 4, 0, 1, 7, 5, 3, 2, 2, 7, 3, 0, 6, 5, 7, 3, 5, 5, 6, 3, 8, 1, 0, 2, 4, 2, 1, 7, 5, 5, 2, 4, 4, 2, 3, 5, 1, 6, 5, 6, 2, 1, 8, 5, 7, 0, 4, 6, 3, 0, 5, 6, 8, 7, 7, 5, 0, 2, 0, 5, 0, 6, 5, 0, 6, 8, 4]
irb(main):014:0> array.all?(3)
=> false
irb(main):015:0> array.my_all?(3)
Traceback (most recent call last):
        5: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        4: from (irb):15
        3: from my_enum.rb:64:in `my_all?'
        2: from my_enum.rb:20:in `my_each'
        1: from my_enum.rb:65:in `block in my_all?'
NoMethodError (undefined method `match?' for 6:Integer)
irb(main):016:0> 
```

File: my_enum.rb [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+      my_each do |i|
+        return true if yield(i)
+      end
+    else
+      my_each do |i|
+        return true if i
+      end
+    end
+    false
+  end
```
Comment (maelfosso) [inline my_enum.rb]:
```ruby
irb(main):019:0> array
=> [6, 5, 0, 7, 8, 7, 8, 6, 4, 0, 5, 1, 7, 6, 4, 4, 7, 2, 1, 0, 2, 4, 8, 8, 0, 7, 3, 8, 2, 2, 2, 1, 2, 8, 4, 0, 1, 7, 5, 3, 2, 2, 7, 3, 0, 6, 5, 7, 3, 5, 5, 6, 3, 8, 1, 0, 2, 4, 2, 1, 7, 5, 5, 2, 4, 4, 2, 3, 5, 1, 6, 5, 6, 2, 1, 8, 5, 7, 0, 4, 6, 3, 0, 5, 6, 8, 7, 7, 5, 0, 2, 0, 5, 0, 6, 5, 0, 6, 8, 4]
irb(main):020:0> array.any?(Integer)
=> true
irb(main):021:0> array.my_any?(Integer)
Traceback (most recent call last):
        5: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        4: from (irb):21
        3: from my_enum.rb:81:in `my_any?'
        2: from my_enum.rb:20:in `my_each'
        1: from my_enum.rb:82:in `block in my_any?'
NoMethodError (undefined method `match?' for 6:Integer)
irb(main):022:0> 
```

File: my_enum.rb [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+  def my_none?(pattern = nil, &block)
+    !my_all?(pattern, &block)
+  end
```
Comment (maelfosso) [inline my_enum.rb]:
```ruby
irb(main):024:0> true_array = [nil, false, true, []]
=> [nil, false, true, []]
irb(main):025:0> true_array.none?
=> false
irb(main):026:0> true_array.my_none?
=> true
irb(main):027:0> 
```

Comment (maelfosso) [reply my_enum.rb]:
```ruby
irb(main):029:0> array
=> [6, 5, 0, 7, 8, 7, 8, 6, 4, 0, 5, 1, 7, 6, 4, 4, 7, 2, 1, 0, 2, 4, 8, 8, 0, 7, 3, 8, 2, 2, 2, 1, 2, 8, 4, 0, 1, 7, 5, 3, 2, 2, 7, 3, 0, 6, 5, 7, 3, 5, 5, 6, 3, 8, 1, 0, 2, 4, 2, 1, 7, 5, 5, 2, 4, 4, 2, 3, 5, 1, 6, 5, 6, 2, 1, 8, 5, 7, 0, 4, 6, 3, 0, 5, 6, 8, 7, 7, 5, 0, 2, 0, 5, 0, 6, 5, 0, 6, 8, 4]
irb(main):030:0> array.none?(String)
=> true
irb(main):031:0> array.my_none?(String)
Traceback (most recent call last):
        6: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        5: from (irb):31
        4: from my_enum.rb:97:in `my_none?'
        3: from my_enum.rb:64:in `my_all?'
        2: from my_enum.rb:20:in `my_each'
        1: from my_enum.rb:65:in `block in my_all?'
NoMethodError (undefined method `match?' for 6:Integer)
irb(main):032:0> 
```

Comment (maelfosso) [reply my_enum.rb]:
```ruby
irb(main):035:0> array
=> [6, 5, 0, 7, 8, 7, 8, 6, 4, 0, 5, 1, 7, 6, 4, 4, 7, 2, 1, 0, 2, 4, 8, 8, 0, 7, 3, 8, 2, 2, 2, 1, 2, 8, 4, 0, 1, 7, 5, 3, 2, 2, 7, 3, 0, 6, 5, 7, 3, 5, 5, 6, 3, 8, 1, 0, 2, 4, 2, 1, 7, 5, 5, 2, 4, 4, 2, 3, 5, 1, 6, 5, 6, 2, 1, 8, 5, 7, 0, 4, 6, 3, 0, 5, 6, 8, 7, 7, 5, 0, 2, 0, 5, 0, 6, 5, 0, 6, 8, 4]
irb(main):036:0> array.none?(5)
=> false
irb(main):037:0> array.my_none?(5)
Traceback (most recent call last):
        6: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        5: from (irb):37
        4: from my_enum.rb:97:in `my_none?'
        3: from my_enum.rb:64:in `my_all?'
        2: from my_enum.rb:20:in `my_each'
        1: from my_enum.rb:65:in `block in my_all?'
NoMethodError (undefined method `match?' for 6:Integer)
irb(main):038:0> 
```

File: my_enum.rb [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    new_object.my_each do |item|
+      if block_given?
+        memo = yield(memo, item)
+      else
+        sym = arg1.is_a?(Symbol) ? arg1 : arg2
+        memo = memo.send(sym, item)
+      end
+    end
+    memo
+  end
```
Comment (maelfosso) [inline my_enum.rb]:
```ruby
irb(main):041:0> range = Range.new(5, 50)
=> 5..50
irb(main):042:0> range.inject(4) { |prod, n| prod * n }
=> 5069015533618896340602101361010794807396273594826752000000000000
irb(main):043:0> range.my_inject(4) { |prod, n| prod * n }
Traceback (most recent call last):
        4: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        3: from (irb):43
        2: from my_enum.rb:120:in `my_inject'
        1: from my_enum.rb:19:in `my_each'
NameError (undefined local variable or method `length' for 5..50:Range)
irb(main):044:0>
```

Comment (maelfosso) [reply my_enum.rb]:
```ruby
irb(main):047:0> range.inject(2, :*)
=> 2534507766809448170301050680505397403698136797413376000000000000
irb(main):048:0> range.my_inject(2, :*)
Traceback (most recent call last):
        4: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        3: from (irb):48
        2: from my_enum.rb:120:in `my_inject'
        1: from my_enum.rb:19:in `my_each'
NameError (undefined local variable or method `length' for 5..50:Range)
irb(main):049:0>
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
Hello @camfilho 

Seriously, this is amazing !!! Awesome !!! :clap: For a first try you did a really really nice job :+1: 
However, there are some bugs you need to fix. Please, read the comment below to know them.

Also, writing the tests is not a part of that project. You must finish this first, test is another project.

Please, kindly make the changes and submit a new code review request.
Thanks.

**Mael FOSSO**
[Github](https://github.com) | [Twitter](https://twitter.com) | [LinkedIn](https://www.linkedin.com/in/mael-elvis-fosso-650b6346/)

**Review by stratospherique — APPROVED** [review]
### Impressive work, congrats :clap: All methods are working properly :heavy_check_mark: The project is approved and you may merge to development :+1: :+1: :+1: 

Keep up the good work :muscle: 
_**Ahmed Mahfoudh**_

**Pull request merged**